src/charm.py

```diff
@@ -35,83 +83,577 @@ def __init__(self, *args: typing.Any):
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self._container = self.unit.get_container("opencti")
+        self._opensearch = self._register_opensearch()
+        self._redis = self._register_redis()
+        self._rabbitmq = self._register_rabbitmq()
+        self._s3 = self._register_s3()
+        self._ingress = self._register_ingress()
+        self._log_forwarder = LogForwarder(self)
+        self._grafana_dashboards = GrafanaDashboardProvider(self)
+        self._metrics_endpoint = MetricsEndpointProvider(
+            self,
+            jobs=[
+                {
+                    "job_name": "opencti_metrics",
+                    "static_configs": [{"targets": ["*:14269"]}],
+                }
+            ],
+        )
+        self.framework.observe(self.on.config_changed, self._reconcile)
+        self.framework.observe(self.on.upgrade_charm, self._reconcile)
+        self.framework.observe(self.on.update_status, self._reconcile)
+        self.framework.observe(self.on.secret_changed, self._reconcile)
+        self.framework.observe(self.on.opencti_pebble_ready, self._reconcile)
+        self.framework.observe(self.on.opencti_peer_relation_created, self._reconcile)
+        self.framework.observe(self.on.opencti_peer_relation_changed, self._reconcile)
+        self.framework.observe(self.on.opencti_peer_relation_departed, self._reconcile)
+        self.framework.observe(
+            self.on["opencti"].pebble_custom_notice, self._on_pebble_custom_notice
+        )
 
-    def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent) -> None:
-        """Define and start a workload using the Pebble API.
+    def _register_opensearch(self) -> OpenSearchRequires:
+        """Create OpenSearchRequires instance and register related event handlers.
 
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
+        Returns:
+            The OpenSearchRequires instance.
 
-        Learn more about interacting with Pebble at at https://juju.is/docs/sdk/pebble.
+        Raises:
+            RuntimeError: If the charm is named 'x-opencti'
+        """
+        if self.app.name == "x-opencti":
+            raise RuntimeError("charm cannot be named 'x-opencti'")
+        opensearch = OpenSearchRequires(
+            self,
+            relation_name="opensearch-client",
+            # suppress the OpenSearch charm from creating the index
+            # use the name x-opencti so OpenSearch will create an index named 'x-opencti'
+            # which shouldn't interfere with the OpenCTI (index prefix is the charm app name)
+            # hope nobody names the charm app 'x-opencti'
```

Comment:
This is not just for one charm application but for all charm applications that connect to the same OpenSearch charm. Therefore, I think a fixed name might have advantages here to avoid collisions among multiple charms.